File: 3. Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        max_sub = 0
        dct = {}
        i = 0
        for j in range(0, len(s)):
        	if s[j] in dct:
        		i = max(dct[s[j]], i)
        	max_sub = max(max_sub, j - i + 1)
        	dct[s[j]] = j + 1
        return max_sub
        

File: 3. Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const length = s.length;
    var max = 0;
    var map = {};
    var i = 0;
    for (var j = 0; j < length; ++j) {
        if (s[j] in map) {
            i = Math.max(i, map[s[j]]);
        }
        max = Math.max(j - i + 1, max);
        map[s[j]] = j + 1;
    }
    return max;
};
